File: shared/src/commonMain/kotlin/com/eter/undiamas/core/data/firebase/FirestoreMappers.kt
package com.eter.undiamas.core.data.firebase

import com.eter.undiamas.core.domain.model.AiMessage
import com.eter.undiamas.core.domain.model.AiMessageRole
import com.eter.undiamas.core.domain.model.CheckInEntry
import com.eter.undiamas.core.domain.model.Mood
import com.eter.undiamas.core.domain.model.MoodEntry
import com.eter.undiamas.core.domain.model.RiskLevel
import com.eter.undiamas.core.domain.model.SupportRole
import com.eter.undiamas.core.domain.model.Trigger
import com.eter.undiamas.core.domain.model.TrustedContact
import com.eter.undiamas.core.domain.model.UserProfile
import com.eter.undiamas.features.diario.domain.DiaryEntry
import kotlin.time.Instant

fun UserProfile.toUsuarioDoc(): UsuarioDoc {
    val contactos = listOfNotNull(trustedContact) + supportNetwork
    return UsuarioDoc(
        alias = displayName,
        fechaInicioSobriedad = sobrietyStartDate.epochSeconds,
        recordRachaSegundos = recordStreakSeconds,
        gastoDiarioEstimado = previousDailyExpense,
        contactosEmergencia = contactos.map { it.toDoc() },
        porQuePersonal = personalWhy,
    )
}

fun UsuarioDoc.toUserProfile(uid: String, existing: UserProfile? = null): UserProfile {
    val contactos = contactosEmergencia.map { it.toDomain() }
    return UserProfile(
        userId = uid,
        displayName = alias,
        sobrietyStartDate = Instant.fromEpochSeconds(fechaInicioSobriedad),
        recordStreakSeconds = recordRachaSegundos,
        previousDailyExpense = gastoDiarioEstimado,
        trustedContact = contactos.firstOrNull(),
        supportNetwork = contactos.drop(1),
        personalWhy = porQuePersonal,
        savingsGoal = existing?.savingsGoal,
    )
}

private fun TrustedContact.toDoc() = ContactoEmergenciaDoc(nombre = name, telefono = phone, rol = role.name)

private fun ContactoEmergenciaDoc.toDomain() = TrustedContact(
    name = nombre,
    phone = telefono,
    role = runCatching { SupportRole.valueOf(rol) }.getOrDefault(SupportRole.FAMILIAR),
)

fun CheckInEntry.toCheckInDoc(): CheckInDoc = CheckInDoc(
    nivelCraving = urgeIntensity,
    estadoAnimo = answers["estado_animo"] ?: "",
    gatillos = triggers.map { it.name },
    nivelRiesgo = riskLevel.name,
    fechaHora = answeredAt.epochSeconds,
    nota = note,
    respuestas = answers,
)

fun CheckInDoc.toCheckInEntry(id: String, uid: String): CheckInEntry = CheckInEntry(
    id = id,
    userId = uid,
    answeredAt = Instant.fromEpochSeconds(fechaHora),
    answers = respuestas.ifEmpty { if (estadoAnimo.isBlank()) emptyMap() else mapOf("estado_animo" to estadoAnimo) },
    riskLevel = runCatching { RiskLevel.valueOf(nivelRiesgo) }.getOrDefault(RiskLevel.VERDE),
    triggers = gatillos.mapNotNull { runCatching { Trigger.valueOf(it) }.getOrNull() }.toSet(),
    urgeIntensity = nivelCraving,
    note = nota,
)

fun DiaryEntry.toDiaryEntryDoc(): DiaryEntryDoc = DiaryEntryDoc(contenido = text, fecha = createdAt.epochSeconds)

fun DiaryEntryDoc.toDiaryEntry(id: String, uid: String): DiaryEntry = DiaryEntry(
    id = id,
    userId = uid,
    createdAt = Instant.fromEpochSeconds(fecha),
    text = contenido,
)

fun MoodEntry.toMoodEntryDoc(): MoodEntryDoc = MoodEntryDoc(animo = mood.name, fecha = registeredAt.epochSeconds)

fun MoodEntryDoc.toMoodEntry(id: String, uid: String): MoodEntry = MoodEntry(
    id = id,
    userId = uid,
    mood = runCatching { Mood.valueOf(animo) }.getOrDefault(Mood.NEUTRAL),
    registeredAt = Instant.fromEpochSeconds(fecha),
)

fun AiMessage.toAiMessageDoc(): AiMessageDoc = AiMessageDoc(
    rol = role.name,
    contenido = content,
    nivelRiesgoContexto = riskLevelContext?.name,
    fecha = sentAt.epochSeconds,
)

fun AiMessageDoc.toAiMessage(id: String, uid: String): AiMessage = AiMessage(
    id = id,
    userId = uid,
    role = runCatching { AiMessageRole.valueOf(rol) }.getOrDefault(AiMessageRole.ASISTENTE),
    content = contenido,
    riskLevelContext = nivelRiesgoContexto?.let { runCatching { RiskLevel.valueOf(it) }.getOrNull() },
    sentAt = Instant.fromEpochSeconds(fecha),
)

File: functions/src/firestoreMappers.ts
import { AiMessageRecord, CheckInRecord, RiskLevel, TrustedContactRecord, UserProfileRecord } from "./types";

/** Forma real de cada entrada de `/usuarios/{uid}` en `contactosEmergencia[]`. */
interface ContactoEmergenciaDoc {
  nombre: string;
  telefono: string;
  rol: string;
}

/** Forma real del documento `/usuarios/{uid}` en Firestore (espejo de UsuarioDoc en Kotlin). */
export interface UsuarioDoc {
  alias?: string;
  fechaInicioSobriedad?: number;
  recordRachaSegundos?: number;
  gastoDiarioEstimado?: number;
  contactosEmergencia?: ContactoEmergenciaDoc[];
  fcmToken?: string;
}

/** Forma real de cada documento en `/usuarios/{uid}/checkins` (espejo de CheckInDoc en Kotlin). */
export interface CheckInDoc {
  nivelCraving?: number;
  estadoAnimo?: string;
  gatillos?: string[];
  nivelRiesgo?: RiskLevel;
  fechaHora?: number;
  nota?: string;
}

/** Forma real de cada documento en `/usuarios/{uid}/mensajesIA` (espejo de AiMessageDoc en Kotlin). */
export interface AiMessageDoc {
  rol?: string;
  contenido?: string;
  nivelRiesgoContexto?: RiskLevel;
  fecha?: number;
}

export function toUserProfileRecord(uid: string, doc: UsuarioDoc): UserProfileRecord {
  const primero = doc.contactosEmergencia?.[0];
  const trustedContact: TrustedContactRecord | undefined = primero
    ? { name: primero.nombre, phone: primero.telefono, role: primero.rol }
    : undefined;
  return { userId: uid, fcmToken: doc.fcmToken, trustedContact };
}

export function toCheckInRecord(id: string, uid: string, doc: CheckInDoc): CheckInRecord {
  return {
    id,
    userId: uid,
    riskLevel: doc.nivelRiesgo ?? "VERDE",
    answeredAt: new Date((doc.fechaHora ?? 0) * 1000).toISOString(),
  };
}

export function toAiMessageRecord(id: string, uid: string, doc: AiMessageDoc): AiMessageRecord {
  return {
    id,
    userId: uid,
    riskLevelContext: doc.nivelRiesgoContexto,
    sentAt: new Date((doc.fecha ?? 0) * 1000).toISOString(),
  };
}
